### PR TITLE
advancedtls: populate verified chains when using custom buildVerifyFunc

### PR DESCRIPTION
**Pull request by mudhireddy**
This request is to address the issue opened https://github.com/grpc/grpc-go/issues/7176

Basically this patch will set the verifiedChains correctly when advancedtls package is used with buildVerifyFunc() and applications need to access TLSInfo.State.VerifiedChains in their logic.

RELEASE NOTES: none

### CODE REVIEW

**Comment by mudhireddy** [comment]
My local testing shows it ok though  

grpc-go/test/xds @ramesh-937457-0.sjc% go test
PASS
ok      google.golang.org/grpc/test/xds 11.131s


Is there a way to trigger the re-test of it?

**Comment by mudhireddy** [comment]
@purnesh42H  any documentation on who and how to add LABEL, MILESTONE so that all checks on this pull request will pass ? right now Mergeable is failing for missing LABEL, MILESTONE

**Comment by purnesh42H** [comment]
@mudhireddy I have assigned it to myself. I will discuss with other maintainers and update you back. Thanks!

File: security/advancedtls/advancedtls.go
Comment (gtcooke94) [inline security/advancedtls/advancedtls.go:621]: We would not want to do this until right before the `return nil` on line 643, otherwise we might set these as verified chains then return an error if we find they are revoked by a CRL. A custom verification func that was supplied could fail as well.

**Review by gtcooke94 — COMMENTED** [review]
Thanks for the PR! In general looks good, one big comment about where we set the value.
I'm reasoning through a few things on the overall code flow to make sure this wouldn't have any risk of leaking cert chains around if someone used the same credential with multiple connections - I'll update here when I'm done looking through that.

In the meanwhile, could you add tests that validate the expected behaviors here?
Also just curious, what is your use case for this?

File: security/advancedtls/advancedtls.go
Comment (mudhireddy) [reply security/advancedtls/advancedtls.go]: Thanks for the review. addressed it in the next patch

File: security/advancedtls/advancedtls.go
Comment (mudhireddy) [reply security/advancedtls/advancedtls.go:564]: done

**Comment by mudhireddy** [comment]
> Thanks for the PR! In general looks good, one big comment about where we set the value. I'm reasoning through a few things on the overall code flow to make sure this wouldn't have any risk of leaking cert chains around if someone used the same credential with multiple connections - I'll update here when I'm done looking through that.
> 
> In the meanwhile, could you add tests that validate the expected behaviors here? Also just curious, what is your use case for this?

Added tests. Reg use case we use the CommonName details from VerifiedChains

**Review by gtcooke94 — DISMISSED** [review]
This looks good, thanks again for the PR!
We'll get one more reviewer to check it out

**Comment by gtcooke94** [comment]
@dfawley can you review this or assign to another go team member?

File: security/advancedtls/advancedtls.go
Comment (dfawley) [inline security/advancedtls/advancedtls.go:512]: This comment seems to add no value.  Please remove it.

File: security/advancedtls/advancedtls.go
Comment (dfawley) [inline security/advancedtls/advancedtls.go:533]: As above.

File: security/advancedtls/advancedtls.go
Comment (dfawley) [inline security/advancedtls/advancedtls.go:567]: A pointer to a slice of slices of pointers to x509 certs?

Would it make more sense to return a second value that is `[][]*x509.Certificate` instead?

Also should we define a type / type alias for `[][]*x509.Certificate`?

Comment (mudhireddy) [reply security/advancedtls/advancedtls.go:565]: to be honest, I couldn't unpack all the syntactic complexity around that to make it work. Any help would be appreciated.

Comment (mudhireddy) [reply security/advancedtls/advancedtls.go:565]: @dfawley 

   cfg := credinternal.CloneTLSConfig(c.config)
   peerVerifiedChains := [][]*x509.Certificate{}
   cfg.VerifyPeerCertificate = buildVerifyFunc(c, "", rawConn, &peerVerifiedChains)
   conn := tls.Server(rawConn, cfg)
   if err := conn.Handshake(); err != nil {
      conn.Close()
      return nil, nil, err
   }

buildVerifyFunc() is returning a function that gets stored in cfg.VerifyPeerCertificate and later gets called as part of conn.Handshake(). Given this structure returning a second parameter would mean that we have to modify the prevalent use of the type "func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error" and that quickly becomes a big change in itself. 

Reg "type / type alias for [][]*x509.Certificate?" I can define it and replace in advancedtls.go where possible, does this name sound reasonable? (sorry, trying to minimize back and forth and want to close this soon)

type CertChains [][]*x509.Certificate

**Comment by mudhireddy** [comment]
@gtcooke94, can I please get approval for this PR to merge ?

File: security/advancedtls/advancedtls.go
Comment (dfawley) [reply security/advancedtls/advancedtls.go]: @gtcooke94 - thoughts?

File: security/advancedtls/advancedtls.go
Comment (gtcooke94) [inline security/advancedtls/advancedtls.go:38]: Can we call it `CertificateChains`?

**Review by gtcooke94 — DISMISSED** [review]
Giving the type a name makes the code a little more readable, I think it's a good change

**Review by gtcooke94 — COMMENTED** [review]
Sorry meant to not approve yet because I don't want the shortened name

File: security/advancedtls/advancedtls.go
Comment (gtcooke94) [reply security/advancedtls/advancedtls.go]: We don't really have too much flexibility in adding return values - we build this verification func then set crypto/tls's `VerifyPeerCertificate` to it https://pkg.go.dev/crypto/tls

File: security/advancedtls/advancedtls.go
Comment (mudhireddy) [reply security/advancedtls/advancedtls.go:38]: sure, let me fix that

**Review by gtcooke94 — APPROVED** [review]
I think this is looking good - it would be nice to be able to replace and use `CertificateChains` everywhere, but we could potentially just do that in a separate PR

**Review by dfawley — APPROVED** [review]

File: security/advancedtls/advancedtls.go
Comment (gtcooke94) [reply security/advancedtls/advancedtls.go]: As far as I can tell this doesn't involve anything async? Instead of passing the variable directly we could do something with channels? But given the constraints of the callback structure, I think something like this is how we'll have to get it out unfortunately.

**Comment by mudhireddy** [comment]
Thank you guys for the review and approval. 

@gtcooke94, I see that it is still open, is it going to auto merge or do I need to do anything for it to merge ?

File: security/advancedtls/advancedtls.go
Comment (arjan-bal) [inline security/advancedtls/advancedtls.go:38]: Since this is an exported type, it should have a doc string.

**Pull request merged**
